clients/pkg/logentry/stages/limit.go

```diff
@@ -0,0 +1,202 @@
+package stages
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"regexp"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/time/rate"
+)
+
+const (
+	ErrLimitStageEmptyConfig        = "limit stage config must contain at least one of `source`, `expression`, `older_than` or `longer_than`"
+	ErrLimitStageInvalidConfig      = "limit stage config error, `value` and `expression` cannot both be defined at the same time."
+	ErrLimitStageInvalidRegex       = "limit stage regex compilation error: %v"
+	ErrLimitStageInvalidRateOrBurst = "limit stage failed to parse rate or burst"
+)
+
+var ratelimitDropReason = "ratelimit_drop_stage"
+
+type LimitConfig struct {
+	Source     *string `mapstructure:"source"`
+	Rate       float64 `mapstructure:"rate"`
+	Burst      int     `mapstructure:"burst"`
+	Drop       bool    `mapstructure:"drop"`
+	Value      *string `mapstructure:"value"`
+	Expression *string `mapstructure:"expression"`
+	regex      *regexp.Regexp
+}
+
+func newLimitStage(logger log.Logger, config interface{}, registerer prometheus.Registerer) (Stage, error) {
+	cfg := &LimitConfig{}
+
+	err := mapstructure.WeakDecode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	err = validateLimitConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &limitStage{
+		logger:      log.With(logger, "component", "stage", "type", "limit"),
+		cfg:         cfg,
+		dropCount:   getDropCountMetric(registerer),
+		rateLimiter: rate.NewLimiter(rate.Limit(cfg.Rate), cfg.Burst),
+	}
+	return r, nil
+}
+
+func validateLimitConfig(cfg *LimitConfig) error {
+	if cfg == nil ||
+		(cfg.Source == nil && cfg.Expression == nil) {
+		return errors.New(ErrLimitStageEmptyConfig)
+	}
+	if cfg.Value != nil && cfg.Expression != nil {
+		return errors.New(ErrLimitStageInvalidConfig)
+	}
+	if cfg.Expression != nil {
+		expr, err := regexp.Compile(*cfg.Expression)
+		if err != nil {
+			return errors.Errorf(ErrLimitStageInvalidRegex, err)
+		}
+		cfg.regex = expr
+	}
+	if cfg.Rate <= 0 || cfg.Burst <= 0 {
+		return errors.Errorf(ErrLimitStageInvalidRateOrBurst)
+	}
+	return nil
+}
+
+// limitStage applies Label matchers to determine if the include stages should be run
+type limitStage struct {
+	logger      log.Logger
+	cfg         *LimitConfig
+	rateLimiter *rate.Limiter
+	dropCount   *prometheus.CounterVec
+}
+
+func (m *limitStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+		for e := range in {
+			if !m.shouldThrottle(e) {
+				out <- e
+				continue
+			}
```

Comment:
I would not expect any side-effects from a function named `shouldThrottle` but only an evaluation whether throttling needs to be applied. What do you think about splitting the check and the actual throttling logic into separate functions?